app/forms/idp/create.tsx

```diff
@@ -111,8 +117,15 @@ export function CreateIdpSideModalForm() {
         name="acsUrl"
         label="ACS URL"
         description="Service provider endpoint for the IdP to send the SAML response"
+        // When creating a SAML identity provider connection, the ACS URL that the user enters
+        // should always be of the form: http(s)://<silo>.sys.<subdomain>/login/<silo>/saml/<name>
+        // where <silo> is the Silo name, <subdomain> is the subdomain assigned to the rack,
+        // and <name> is the name of the IdP connection
+        value={`https://${silo}.sys.${subdomain}/login/${silo}/saml/${name}`}
```

Comment:
Looks like I was wrong about this working fine. It works fine to display it in the input, but it doesn't appear to be syncing with the RHF form state like I thought it would. Not sure if there's a way to make it do that or maybe we are actually stuck with the `useEffect`. I thought about using an `onChange` on the name field, which is usually preferable, but at least as it is implemented now, passing `onChange` to `NameField` breaks it by overrides the `onChange` that hooks the field up to the RHF state.

<img width="493" alt="image" src="https://github.com/user-attachments/assets/2ec91700-9ff5-46cd-801d-e68b956269f1">

Reply:
Ah, crumbs; well, easily revertable. I guess we know now why we set the form the way we do!

Reply:
It's not that hard to make `onChange` stop being such a jerk, but it's not a complete fix — `calling `setValue` in `onChange` does work, but you also have to set the initial value. I don't love it, just an idea.

```diff
diff --git a/app/components/form/fields/TextField.tsx b/app/components/form/fields/TextField.tsx
index 9163a02e..ab576c89 100644
--- a/app/components/form/fields/TextField.tsx
+++ b/app/components/form/fields/TextField.tsx
@@ -118,7 +118,11 @@ export const TextFieldInner = <
         type={type}
         error={!!error}
         aria-labelledby={`${id}-label ${id}-help-text`}
-        onChange={(e) => onChange(transform ? transform(e.target.value) : e.target.value)}
+        onChange={(value) => {
+          const transformed = transform ? transform(value) : value
+          onChange(transformed)
+          props.onChange?.(transformed)
+        }}
         {...fieldRest}
         {...props}
       />
diff --git a/app/forms/idp/create.tsx b/app/forms/idp/create.tsx
index 601abc65..4d1e44e5 100644
--- a/app/forms/idp/create.tsx
+++ b/app/forms/idp/create.tsx
@@ -22,10 +22,8 @@ import { pb } from '~/util/path-builder'
 
 import { MetadataSourceField, type IdpCreateFormValues } from './shared'
 
-const host = window.document.location.host.toString()
-const subdomain = host.includes('localhost')
-  ? 'r3.oxide-preview.com'
-  : host.split('.').slice(2).join('.')
+const host = window.location.hostname
+const subdomain = host === 'localhost' ? 'localhost' : host.split('.').slice(2).join('.')
 
 const defaultValues: IdpCreateFormValues = {
   type: 'saml',
@@ -63,8 +61,16 @@ export function CreateIdpSideModalForm() {
     },
   })
 
-  const form = useForm({ defaultValues })
-  const name = form.watch('name')
+  function getAcsUrl(name: string) {
+    return `https://${silo}.sys.${subdomain}/login/${silo}/saml/${name}`
+  }
+
+  const form = useForm({
+    defaultValues: {
+      ...defaultValues,
+      acsUrl: getAcsUrl(defaultValues.name),
+    },
+  })
 
   return (
     <SideModalForm
@@ -111,17 +117,20 @@ export function CreateIdpSideModalForm() {
       submitError={createIdp.error}
       submitLabel="Create provider"
     >
-      <NameField name="name" control={form.control} />
+      <NameField
+        name="name"
+        control={form.control}
+        // When creating a SAML identity provider connection, the ACS URL that the user enters
+        // should always be of the form: http(s)://<silo>.sys.<subdomain>/login/<silo>/saml/<name>
+        // where <silo> is the Silo name, <subdomain> is the subdomain assigned to the rack,
+        // and <name> is the name of the IdP connection
+        onChange={(name) => form.setValue('acsUrl', getAcsUrl(name))}
+      />
       <DescriptionField name="description" control={form.control} required />
       <TextField
         name="acsUrl"
         label="ACS URL"
         description="Service provider endpoint for the IdP to send the SAML response"
-        // When creating a SAML identity provider connection, the ACS URL that the user enters
-        // should always be of the form: http(s)://<silo>.sys.<subdomain>/login/<silo>/saml/<name>
-        // where <silo> is the Silo name, <subdomain> is the subdomain assigned to the rack,
-        // and <name> is the name of the IdP connection
-        value={`https://${silo}.sys.${subdomain}/login/${silo}/saml/${name}`}
         required
         disabled
         control={form.control}
@@ -135,6 +144,7 @@ export function CreateIdpSideModalForm() {
         description="Service provider endpoint for log out requests"
         required
         control={form.control}
+        defaultValue={`https://${silo}.sys.${subdomain}/login/${silo}/saml/`}
       />
       {/* TODO: help text */}
       <TextField
diff --git a/app/ui/lib/TextInput.tsx b/app/ui/lib/TextInput.tsx
index bb635b50..ab2c8bf4 100644
--- a/app/ui/lib/TextInput.tsx
+++ b/app/ui/lib/TextInput.tsx
@@ -8,6 +8,7 @@
 import { announce } from '@react-aria/live-announcer'
 import cn from 'classnames'
 import React, { useEffect } from 'react'
+import { type Merge } from 'type-fest'
 
 import { CopyToClipboard } from './CopyToClipboard'
 import { Tooltip } from './Tooltip'
@@ -35,14 +36,18 @@ export type TextAreaProps =
 // it makes a bunch of props required that should be optional. Instead we simply
 // take the props of an input field (which are part of the Field props) and
 // manually tack on validate.
-export type TextInputBaseProps = React.ComponentPropsWithRef<'input'> & {
+export type TextInputBaseProps = Merge<
+  React.ComponentPropsWithRef<'input'>,
+  {
     // error is used to style the wrapper, also to put aria-invalid on the input
     error?: boolean
     disabled?: boolean
     className?: string
     fieldClassName?: string
     copyable?: boolean
+    onChange?: (value: string) => void
   }
+>
 
 export const TextInput = React.forwardRef<
   HTMLInputElement,
@@ -58,6 +63,7 @@ export const TextInput = React.forwardRef<
       fieldClassName,
       copyable,
       as: asProp,
+      onChange,
       ...fieldProps
     },
     ref
@@ -77,6 +83,7 @@ export const TextInput = React.forwardRef<
         )}
         aria-invalid={error}
         disabled={disabled}
+        onChange={(e) => onChange?.(e.target.value)}
         {...fieldProps}
       />
     )
```

Reply:
probably better to stick with the `useEffect` for now 😁